mantidproject/mantid
# SANSILLAutoProcess multiple transmissions

**Description of work.**

Added support for multiple transmissions to be processed by SANSILLAutoProcess. The dimensionality of the provided transmission input must be either one or consistent with the dimensionality of the relevant sample (one from absorber, beam, container, sample). In case one transmission numor is provided, that transmission workspace will be applied to all provided numors. When more than one transmission is provided (and is consistent with the number of relevant sample numors), each transmission will be applied to data in the same order as sample numors are processed.



**To test:**
1. Make sure you have SystemTest data with ILL/D11 directory in your user data path.
2. Run the following through the mantidpython or workbench:

```
beams = '1020,947,1088'
containers = '1023,973,1003'
container_tr = '1023,988,988'
beam_tr = '1020,1119,1119'
samples = '1025,975,1005'
sample_tr = '1204,990,990'
thick = 0.1

#mask edges first:
MaskBTP(Instrument='D11', Tube='1-3,253-256')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_vertical')
MaskBTP(Instrument='D11', Pixel='1-3,253-256')
Plus(LHSWorkspace='mask_vertical', RHSWorkspace='D11MaskBTP', OutputWorkspace='edge_masks')
# then, the edges mask can be used as a default mask for all distances and wavelengths
MaskBTP(Instrument='D11', Tube='116-139', Pixel='90-116')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_39m_10A')
MaskBTP(Instrument='D11', Tube='115-140', Pixel='115-140')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_8m_4_6A')
MaskBTP(Instrument='D11', Tube='105-145', Pixel='105-145')
RenameWorkspace(InputWorkspace='D11MaskBTP', OutputWorkspace='mask_1m_4_6A')

# reduce samples
# this also tests that already loaded workspace can be passed instead of a file
LoadNexusProcessed(Filename='sens-lamp.nxs', OutputWorkspace='sens-lamp')
SANSILLAutoProcess(
            SampleRuns=samples,
            BeamRuns=beams,
            ContainerRuns=containers,
            DefaultMaskFile='edge_masks',
            MaskFiles='mask_39m_10A,mask_8m_4_6A,mask_1m_4_6A',
            SensitivityMaps='sens-lamp',
            SampleTransmissionRuns=sample_tr,
            ContainerTransmissionRuns=container_tr,
            TransmissionBeamRuns=beam_tr,
            SampleThickness=thick,
            CalculateResolution='MildnerCarpenter',
            OutputWorkspace='iq_mult_wavelengths',
            BeamRadius='0.05',
            TransmissionBeamRadius=0.05
)
```
3. There should be no issues from the input validation or the processing.
4. Try changing the number of numors in the sample_tr to more or less than 3 entries, observe the input validation feedback: the number of transmission inputs should be either consistent with the number of sample runs or equal to one. 

Fixes #30184

Release notes added.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
